### PR TITLE
Async opportunity search support

### LINKED ISSUE #77: Allow `opportunities` resource search to operate asynchronously
Computing opportunities can take a bit too long to be handled by a synchronous request. In the synchronous request for the sample backend, at least for the Planet API, we in fact encapsulate an asynchronous request and only return once it's finished. Also, if a customer polls results from the async opportunity collection process and only a part of the results have been gathered, it would be good if those results would already be shown, along with information how to get the following ones (basically pagination, but not creating page breaks based on the number of results, but on the time of creation). See https://developers.planet.com/docs/tasking/examples/assured/ for one way to expose this kind of functionality, in particular the "next" field that would display the next set of results (even though unused in the example).

### PR DESCRIPTION
**Pull request by jkeifer**
Closes #77.

### CODE REVIEW

**Comment by jkeifer** [comment]
Right now this doesn't address ordering. It seems like ordering already supports this operation with the statuses outlined, but it might make sense to call out the difference in ordering between sync search and not?

Also, prior discussion indicated that we should keep it simple and not provide a mechanism for clients to request sync vs async, but it seemed to me that a user might care what operation they get because they are not entirely operationally equivalent. So I added in a section about the Prefer header and how to use that to control search behavior. I could be convinced to pull that out, but I also feel more strongly now that similar support should be extended to the order endpoint _OR_ the spec needs to call out exactly how to reconcile the difference between sync and async searches in ordering via the specified statuses. Perhaps this would also necessitate an enumerated status code to indicate searching opportunities failed to return one.

**Comment by jkeifer** [comment]
The openapi docs have not been updated. I'm holding off on that until I get some preliminary feedback.

File: opportunity/README.md [hunk 1]
```diff
@@ -73,12 +86,69 @@ resources below.
...
+
+### Opportunity Search Record
+
+Returned by an async opportunity search. Can also be retrieved directly.
+
+| Field Name | Type                                                                       | Description |
+|------------| -------------------------------------------------------------------------- | ----------- |
+| id         | string                                                                     | **REQUIRED.** Opportunity search record ID. |
+| productId  | string                                                                     | **REQUIRED.** Product identifier. This should be a reference to the [Product](https://github.com/Element84/stapi-spec/blob/main/product/README.md) being searched. |
+| status     | SearchStatus                                                               | **REQUIRED.** The current search status. |
```
Comment (CloudNiner) [inline opportunity/README.md:123]: Are the valid SearchStatus values documented elsewhere?

File: opportunity/README.md [hunk 2]
```diff
@@ -73,12 +86,69 @@ resources below.
...
+Search Record (`/searches/opportunities/{searchRecordId}`).
+
+### Sync vs Async Opportunity Searching
+
+Servers that advertise support for async searching via the conformance class
+must support async searches for all products. That said, servers can still
+default to a sync response for products that support it, if desired.  Clients
+can request sync vs async operation via the HTTP `Prefer` header.
+
+Possible values for the `Prefer` header are `respond-async` or `wait`, where
```
Comment (CloudNiner) [inline opportunity/README.md:149]: Hmmm interesting. Seems like a tradeoff of client complexity for increased provider implementation flexibility? In practice if a client could choose sync or async wouldn't a client always want sync for speed and implementation simplicity if they could get it? What was the use case that led to proposing this workflow?

I do wonder if any provider implementation would actually implement this endpoint in such a way that it could actually respond dynamically with sync or async? Maybe that capability could be sold as a competitive advantage against other providers?

File: opportunity/README.md [hunk 3]
```diff
@@ -73,12 +86,69 @@ resources below.
...
+### Opportunity Search Record
+
+Returned by an async opportunity search. Can also be retrieved directly.
+
+| Field Name | Type                                                                       | Description |
+|------------| -------------------------------------------------------------------------- | ----------- |
+| id         | string                                                                     | **REQUIRED.** Opportunity search record ID. |
+| productId  | string                                                                     | **REQUIRED.** Product identifier. This should be a reference to the [Product](https://github.com/Element84/stapi-spec/blob/main/product/README.md) being searched. |
+| status     | SearchStatus                                                               | **REQUIRED.** The current search status. |
+| links      | \[[Link Object](#link-object)]                                             | List of link objects to resources and related URLs. There **must** be a `rel=opportunities` link to the corresponding opportunity collection when the search is completed. See [Link Object - rel=opportunities](#rel=opportunities) below. |
```
Comment (CloudNiner) [inline opportunity/README.md:124]: suggestion non blocking: Should there also be a `rel=search-record` or equivalent that points clients to the `GET /searches/opportunities/{searchRecordId}` endpoint too?

**Review by CloudNiner — COMMENTED** [review]
Nice to see async being thought about early -- I'd imagine the spec would need to support it for most robust provider implementations.

> Right now this doesn't address ordering. It seems like ordering already supports this operation with the statuses outlined, but it might make sense to call out the difference in ordering between sync search and not?

Could you clarify this a bit? Does the ordering workflow change whether you requested async or sync opportunities? It doesn't seem like it should. Or maybe I'm wildly misunderstanding the comment 😬

**Comment by jkeifer** [comment]
>> Right now this doesn't address ordering. It seems like ordering already supports this operation with the statuses outlined, but it might make sense to call out the difference in ordering between sync search and not?
>
> Could you clarify this a bit? Does the ordering workflow change whether you requested async or sync opportunities? It doesn't seem like it should. Or maybe I'm wildly misunderstanding the comment 😬

@CloudNiner The issue here is that ordering, by taking a search request as the request payload, has to perform an opportunity search to identify a single opportunity to order. If going through the flow of `search opportunities -> pick one -> order it`, the order endpoint gets a search request that is constrained (hopefully) to a single opportunity, but it still has to do that search to identify that specific opportunity.

In other words, the order endpoint has to support requests that might return multiple opportunities, and cannot know when a request maps to a single opportunity without performing a search. Thus it has to do the search and have some ability to identify which to pick if more than one opportunity is returned (though the simple and naive solution is to just take the first returned).

Now there is an exception to what I stated above, namely if opportunities have unique identifiers and the implementation is such that the request can contain a specific opportunity  identifier. Such a case is implementation specific and cannot be counted upon in this spec, however.

But I digress, let's get back to the issue of ordering. I think two of the currently-outline order statuses apply here:

* `received` (indicates order received by provider and it passed format validation)
* `accepted` (indicates order has been accepted)

It's also worth noting that, currently, the Order entity as implemented in stapi-fastapi does not entirely align with the spec. See https://github.com/stapi-spec/stapi-spec/issues/223 for a relevant issue and https://github.com/stapi-spec/stapi-fastapi/blob/main/src/stapi_fastapi/models/order.py#L79 for the current implementation. The summary is that the `Order` model in stapi-fastapi is a geojson feature, and it is expected that it gets its geometry and some properties from the ordered opportunity. In other words, constructing an `Order` currently requires a single opportunity to have been identified. (Now of course implementation does not equal spec, and stapi-fastapi might be wrong or overly specific, but I think it is useful context for consideration here).

So we have an option here. We could always create an Order with the `received` status without identifying an opportunity, but it would not be a valid geojson feature. We could then make it valid and move it to `accepted` once we identify and reserve the opportunity. Doing this would require being able to reliably run the opportunity search in the background after returning the response to the client, which is not necessarily something an implementation might want to support if opportunity searching is only sync.

Alternatively, when sync searching is supported for a product, we could run the search as part of building the Order instance for return to the client. It seems in this case the first status the client would see for the order would be `accepted`; `received` would be skipped. Keeping the opportunity search in the context of preparing the response means that the response is blocked but the implementation does not have to handle background tasks. Thus it seems to me that there is a fundamental difference between sync ordering and async ordering from the requirements imposed upon an implementation (the latter already requires background task support to handle async opportunities requests).

Now in the async search case we could do as I proposed above, creating an invalid Order feature with the `received` status, but upon writing this up I realize the invalid geojson feature is a bit of a problem. Instead, we could specify that a `202 Accepted` status be returned for orders leveraging async searching. If no opportunity could be identified (or some other error happens) the order could be created in the `rejected` status (and geom could come from the search request in this case instead of an identified opportunity). Maybe? I don't know, perhaps this isn't great because listing orders likely couldn't include these orders awaiting an identified opportunity.

So this is a lot and I don't know if it really helps or not. 😅 The point is that ordering requires a specific opportunity to be identified (as far as I can tell), and doing so requires searching. In the case of a totally sync implementation searching within the request lifetime makes the most sense. But with async that's not possible, and thus the behavior of ordering varies depending on whether sync or async opportunity searching is available for the given product.

File: opportunity/README.md (hunk 1 above)
Comment (jkeifer) [reply opportunity/README.md:123]: No, this is a failure on my part. I guess I must have been interrupted and forgot where I was at when I got back to this. I will get those added tomorrow.

File: opportunity/README.md (hunk 2 above)
Comment (jkeifer) [reply opportunity/README.md:149]: While you represent sync as superior, I think there's value to the client in choosing async. With async the search is saved and can be retrieved at a later time to be rerun. The opportunities are also persisted (for some duration, probably not indefinitely) and could be read multiple times. I could imagine a UI client, for example, leveraging async support to allow multiple searches to be executed in parallel, and use the server-side persistence to enable switching between result sets. Additionally, if one product only supports async while others can support sync, it might be easier as a client to standardize on async to simplify the implementation.

> I do wonder if any provider implementation would actually implement this endpoint in such a way that it could actually respond dynamically with sync or async?

Sync is easy to support server side. But if a server has to support async for one product that requires it, then leveraging that async implementation across all other products would be trivial for any clients would prefer async.

File: opportunity/README.md (hunk 3 above)
Comment (jkeifer) [reply opportunity/README.md:124]: I like this idea, I will get it added as a recommendation.

File: opportunity/README.md (hunk 2 above)
Comment (jkeifer) [reply opportunity/README.md:149]: Oh, forgot to mention, this idea is lifted from the OGC Processes API spec (https://docs.ogc.org/is/18-062r2/18-062r2.html#toc32:~:text=requirement%20A).-,7.11.2.3,-.%C2%A0%20Execution%20mode), which itself based its use of the Prefer header on the [Prefer Header for HTTP RFC](https://www.rfc-editor.org/rfc/rfc7240.txt). That RFC outlines the `request-async` and `wait` preferences.

File: README.md [hunk 4]
```diff
@@ -79,73 +97,104 @@ Fields that can be included in the response body for `GET /`.
...
+| `POST /products/{productId}/opportunities`               | Opportunities       | [Opportunity Request](./opportunity/README.md#opportunity-request) | Sync search: [Opportunities Collection](./opportunity/README.md#opportunities-collection); Async search: [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | Explore the opportunities available for a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) |
+| `GET /products/{productId}/opportunities/{collectionId}` | Async Opportunities | -                                                                  | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Get the opportunity collection for an async opportunity search              |
+| `GET /searches/opportunities/`                           | Async Opportunities | -                                                                  | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] |          |
```
Comment (philvarner) [inline README.md:153]: I don't think this is an endpoint, maybe just a copy error?

File: README.md [hunk 5]
```diff
@@ -79,73 +97,104 @@ Fields that can be included in the response body for `GET /`.
...
+| **Name**                    | **Specified in**                       | **Conformance URI**                                  | **Description**                                                                          |
+| --------------------------- | -------------------------------------- | ---------------------------------------------------- | ---------------------------------------------------------------------------------------- |
+| STAPI - Core                | Core                                   | https://stapi.example.com/v0.1.0/core                | Specifies the STAPI Landing page `/`, communicating conformance and available endpoints. |
+| STAPI - Opportunities       | [Opportunities](opportunity/README.md) | https://stapi.example.com/v0.1.0/opportunities       | Enables request of potential tasking opportunities                                       |
+| STAPI - Async Opportunities | [Opportunities](opportunity/README.md) | https://stapi.example.com/v0.1.0/async-opportunities | Enables async request of potential tasking opportunities                                 |
```
Comment (philvarner) [inline README.md:175]: I'd prefer this to be `opportunities-async` so it's more apparently related to `opportunities`

File: README.md (hunk 4 above)
Comment (philvarner) [inline README.md:153]: I'm unclear as to what this endpoint does. Why does it have a trailing slash? and why does it return the same type as /searches/opportunities/{searchRecordId}?

File: opportunity/README.md [hunk 6]
```diff
@@ -73,12 +86,69 @@ resources below.
...
 
 If no Body parameters apply to an Opportunity, use `"body": {}`.
+
+## Async Opportunity Search
+
+- **Conformance URI:** <https://stapi.example.com/v0.1.0/async-opportunities>
+
+STAPI has an optional conformance class providing support for async opportunity
+searches, to accommodate searches for products that require more time than is
+allowable by the duration of an HTTP request. This support requires persisting
```
Comment (philvarner) [inline opportunity/README.md:109]: maybe "allowable" -> "desired" or "typical"

File: opportunity/README.md [hunk 7]
```diff
@@ -73,12 +86,69 @@ resources below.
...
+| Field Name | Type                                                                       | Description |
+|------------| -------------------------------------------------------------------------- | ----------- |
+| id         | string                                                                     | **REQUIRED.** Opportunity search record ID. |
+| productId  | string                                                                     | **REQUIRED.** Product identifier. This should be a reference to the [Product](https://github.com/Element84/stapi-spec/blob/main/product/README.md) being searched. |
+| status     | SearchStatus                                                               | **REQUIRED.** The current search status. |
+| links      | \[[Link Object](#link-object)]                                             | List of link objects to resources and related URLs. There **must** be a `rel=opportunities` link to the corresponding opportunity collection when the search is completed. See [Link Object - rel=opportunities](#rel=opportunities) below. |
+
+#### rel=opportunities
+
+This Link object provides the means of retireving the search results when the
```
Comment (philvarner) [inline opportunity/README.md:128]:
```suggestion
This Link object provides the means of retrieving the search results when the
```

File: opportunity/README.md [hunk 8]
```diff
@@ -73,12 +86,69 @@ resources below.
...
+search is completed. That is, it should include the equivalent of `GET
+/products/{productId}/opportunities/{collectionId}` where `productId` is the
```
Comment (philvarner) [inline opportunity/README.md:130]: maybe rename `collectionId` to `opportunitiesCollectionId` -- whenever I see "collection", I always think of a STAC Collection

File: opportunity/README.md (hunk 2 above)
Comment (CloudNiner) [reply opportunity/README.md:149]: Yeah, fair enough. For the example:

> UI client leveraging async support to allow multiple searches to be executed in parallel

The UI client could do that for sync requests and cache the state locally. It's just moving the state complexity from one place to another. Neither feels necessarily better.

I don't really feel strongly about it in any case, was just looking to explore use cases, so carry on! I imagine this is the type of thing that the spec would get a lot of feedback back about when implementations get built. Perhaps a discussion topic for the next sprint?

**Comment by CloudNiner** [comment]
> Thus it has to do the search and have some ability to identify which to pick if more than one opportunity is returned (though the simple and naive solution is to just take the first returned).

Yup. That's intentional. The provider implementing Orders for any given Product can determine how to best choose which of the opportunities is the best fit. That gives all sorts of benefits like allowing schedule optimization behind the scenes. Or shifting the collection after ordering when other orders come in with higher precedence, similar constraints, etc.

> In other words, constructing an `Order` currently requires a single opportunity to have been identified. 

Ah, I think this is where the thread gets off target. I believe in lots of cases this statement isn't true.

Some examples... if the provider implementation needs time to figure out a matching specific opportunity that implementation can return Order Feature values that match whatever was submitted without further constraining. In many cases for many providers, the actual opportunity collected could and probably would change over time. In addition, there's lots of providers that if you give them a set of constraints when ordering will only ever tell you "Yes" or "No" and never be more specific than that. The customer never sees the specific opportunity selected until they get their collected assets delivered way later. Or, a provider may have some heuristic about accepting certain constraints that its <100% sure it could fulfill, but accept anyway and figure it out or REJECT later, if the constraints are sufficiently wide.

I think all of that simplifies the rest of your comments? Hopefully? Regardless of sync or async, an Order Feature at minimum can be constructed that matches the input constraints submitted by the user. Then the implementation moves to ACCEPTED (Umbra calls it SCHEDULED) when a matching opportunity is found and scheduled. Up to the provider whether to update the Order feature or indicate in some other way to the STAPI user whenever the specific opportunity changes at that or any other time before collection, if at all.

**Comment by philvarner** [comment]
> Ah, I think this is where the thread gets off target. I believe in lots of cases this statement isn't true.

I agree. I see it as somewhat up to the user to decide what Order constraints they actually apply, and then that relates to a widening or narrowing of future possibilities. For example, a user could search for opportunities over a 2 week interval, and get Opportunities for day 3, day 7, and day 12. They may then re-search with a tighter constraint that only selects day 7, and submit their order -- but with the tradeoff that their order can only be fulfilled on that day, and not earlier on day 3 or later (without having to re-order) on day 12. If they submit the original search order that matches the 3 days (or maybe others as other orders are cancelled), then it's more likely to be fulfilled. I could also see pricing being built into this, so that an order is cheaper if there are more options for how the provider can actually fulfill it.

**Comment by jkeifer** [comment]
>> In other words, constructing an Order currently requires a single opportunity to have been identified.
>
> Ah, I think this is where the thread gets off target. I believe in lots of cases this statement isn't true.

@CloudNiner While I've completely understood the points you make about this, I must admit I had a very different concept of how to deal with these concerns--it has entirely been about shaping opportunities to fit the flexibility required by the vendor, as an order has been to me one-to-one with an opportunity.

Allow me to 🤯 for a moment. Now that I've done that and the fire is out, allow me to say: I stand corrected 🙇. I think I've been conceiving of an Order incorrectly the entire time. I expect I am not the only one, so I took a step back to think about where I went wrong here. Stepping back also involved writing way to many words (sorry in advance), but I managed to compile my thoughts and a proposal into #241.

The short version: I think the data model is potentially misleading, and think we need to explicitly decouple Orders from Opportunities selected to fulfill the former by making defining a data model where an Order has a one-to-many relation with Opportunities. I believe there are several benefits to such a data model, not the least of which is that I believe it would more concretely define what an Order is in such a way as to not allow the confusion that I've been experiencing.

File: opportunity/README.md (hunk 8 above)
Comment (jkeifer) [reply opportunity/README.md:130]: I just didn't want to make an overly-long table line even longer, but you are right, it should be `opportunitiesCollectionId`.

File: README.md (hunk 4 above)
Comment (jkeifer) [reply README.md:153]: It's hard to see in the raw form but this endpoint is returning an array of Opportunity Search Records. This endpoint allows a user to list all their searches in one place. The trailing slash was an error on my part and I will remove it.

File: README.md [hunk 9]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
+| `GET /products/{productId}/opportunities/{opportunityCollectionId}` | Opportunities (Async) | - | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Get the opportunity collection for an async opportunity search |
+| `GET /searches/opportunities` | Opportunities (Async) | - | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] | |
+| `GET /searches/opportunities/{searchRecordId}` | Opportunities (Async) | - | [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | |
+| `GET /searches/opportunities/{searchRecordId}/statuses` | Opportunities (Async) | - | [[Opportunity Search Status](./opportunity/README.md#opportunity-search-status)] | |
+| `POST /searches/opportunities/{searchRecordId}/statuses` | Opportunities (Async) | [Opportunity Search Status](./opportunity/README.md#opportunity-search-status) | - | POST a search status update. Returns an empty 201 on success. |
```
Comment (jkeifer) [inline README.md:156]: The whole statuses piece here borrows heavily from the `order-statuses` extension. However, the spec does not outline an equivalent of this endpoint for order statuses. I am uncertain if this is because it is intended to be an implementation-specific endpoint and not addressable by the spec or if it is an oversight.

If I should remove this endpoint from the spec let me know. Similarly, if we need to add an endpoint to update order statuses let me know that too and I can see about getting a change in for that (or at least an issue).

File: opportunity/README.md [hunk 10]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+### Sync vs Async Opportunity Searching
+
+Servers that advertise support for async searching via the conformance class
+must support async searches for all products. That said, servers can still
+default to a sync response for products that support it, if desired.  Clients
+can request sync vs async operation via the HTTP `Prefer` header.
+
+Possible values for the `Prefer` header are `respond-async` or `wait`, where
+the former requests as async response and the latter requests a sync response.
+The server must always respond with the HTTP `Preference-Applied` header
```
Comment (rkm-ursa) [inline opportunity/README.md:191]: What is the value returned in the `Preference-Applied` header? Is it a boolean or does it mimic the `Prefer` header or is it something else entirely? This should be explicitly laid out.

Comment (swallace-ursa) [reply opportunity/README.md:191]: The RFC outlining this pattern is here: https://www.rfc-editor.org/rfc/rfc7240.txt
I believe the implication is that since multiple `Prefer` headers can be applied to a request, `Preference-Applied` will return the full preference that was applied rather than a boolean.

**Review by swallace-ursa — COMMENTED** [review]
There will likely be variances between vendors on the following areas:
* Data retention periods
* Preferred polling frequency

Each implementation would need to document how these are handled, and the user of these systems would need to handle them on a case-by-case basis. Would it be beneficial to address any of this in the spec so that the handling of the data can be addressed programmatically rather than having vendor-specific implementations? It feels heavy for the spec, but just writing out my thoughts.

----

I've been going back and forth with the usefulness of the `Prefer` header, but I have concluded that there are scenarios where it is in fact beneficial, and it should not induce a burden on a server to handle this.

**Comment by swallace-ursa** [comment]
As I think about this from the perspective of vendor integration, I'm realizing that an ideal system to me would not involve polling for updates at any stage of it, but rather would be event-driven.

An endpoint to check the status (both for tasking and orders) makes sense for debugging, or if an expected event is overdue, but the happy path should be for the user to wait for the vendor to call them back when their data is ready.

Granted, this adds complexity to the spec. Perhaps keeping this as-is, but then also adding a conformance class for webhook support is an approach that could be taken?

**Comment by CloudNiner** [comment]
> As I think about this from the perspective of vendor integration, I'm realizing that an ideal system to me would not involve polling for updates at any stage of it, but rather would be event-driven.
> 
> 
> 
> An endpoint to check the status (both for tasking and orders) makes sense for debugging, or if an expected event is overdue, but the happy path should be for the user to wait for the vendor to call them back when their data is ready.
> 
> 
> 
> Granted, this adds complexity to the spec. Perhaps keeping this as-is, but then also adding a conformance class for webhook support is an approach that could be taken?

👍👍👍 100% yes! I 😭 every time I think about our own API not having an event driven alerting mechanism because it's such a bad experience to poll. 

Probably best as a separate issue though?

**Comment by philvarner** [comment]
> Granted, this adds complexity to the spec. Perhaps keeping this as-is, but then also adding a conformance class for webhook support is an approach that could be taken?

Agree, either a conformance class or as a recommended way to do it that's documented out-of-band.

File: README.md [hunk 11]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
+and before `POST /products/{productId}/orders`. It allows users more
+fine-grained control and selection of available tasking opportunities by
+letting them explore the opportunities which are available for a chosen order
+configuration. The opportunities are represented in a FeatureCollection, with
+order specific attributes and values in the feature properties.
+
+Implementations can choose to support synchronous or asynchronous opportunity
+search requests as needed to support the duration of the operations required to
+find opportunities for a given Product. In cases where the opportunity search
+can be completed within the duration of a regular HTTP request synchronous
```
Comment (philvarner) [inline README.md:117]:
```suggestion
can be completed within the duration of a regular HTTP request, synchronous
```

File: README.md [hunk 12]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
 STAPI supports paging through hypermedia links for the following resources:
+
 - `GET /products`
 - `POST /products/{productId}/opportunities`
+- `GET /products/{productId}/opportunities/{collectionId}`
 - `GET /orders`
 - `GET /orders/{orderId}/statuses`
+- `GET /searches/opportunities/`
 
 The following relation types are available for pagination:
```
Comment (philvarner) [inline README.md:198]:
```suggestion
The following relation types may be available for pagination:
```

File: README.md [hunk 13]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
-| `POST /products/{productId}/orders`          | Core          | [Order Request](./order/README.md#order-request) or any object     | -                                                                            | Order a capture with a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) as defined in the products or a request that was provided through the opportunities endpoint. |
-| `POST /products/{productId}/opportunities`   | Opportunities | [Opportunity Request](./opportunity/README.md#opportunity-request) | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Explore the opportunities available for a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters)                                                                            |
+STAPI follow the modern web API practices of using HTTP Request Methods
```
Comment (philvarner) [inline README.md:132]:
```suggestion
STAPI follows the modern web API practices of using HTTP Request Methods
```

File: README.md [hunk 14]
```diff
@@ -177,12 +228,13 @@ a parameter `page` and is currently on page 2:
 
 The href may contain any arbitrary URL parameter, which is implementation-specific:
 
-- `https://stac-api.example.com/collections/my_collection/items?page=2`
-- `https://stac-api.example.com/collections/my_collection/items?next=8a35eba9c`
-- `https://stac-api.example.com/collections/my_collection/items?token=f32890a0bdb09ac3`
+- `https://stapi.example.com/products?page=2`
+- `https://stapi.example.com/productsnext=8a35eba9c`
```
Comment (philvarner) [inline README.md:232]:
```suggestion
- `https://stapi.example.com/products?next=8a35eba9c`
```

File: opportunity/README.md [hunk 15]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+geospatially.  STAC compliant APIs are required to compute intersection
+operations with the Item's geometry field, not its bbox.
 
 #### Properties Object
 
-Additional metadata fields can be added to the GeoJSON Object Properties that describe the Opportunity in more detail for the user. The only required fields are  `datetime` but it is recommended to add more fields, see [Additional Fields](#additional-fields)
-resources below.
+Additional metadata fields can be added to the GeoJSON Object Properties that
+describe the Opportunity in more detail for the user. The only required fields
+are `datetime` and `product_id but it is recommended to add more fields as
```
Comment (philvarner) [inline opportunity/README.md:104]:
```suggestion
are `datetime` and `product_id` but it is recommended to add more fields as
```

File: opportunity/README.md [hunk 16]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+- in_progress (indicates search is running)
+- failed (indicates search will not be fulfilled for some error reason)
+- canceled (indicates search was canceled for any reason)
+- completed (indicates search has completed successfully and the results can be retrieved)
+
+Providers must support these statuses.
+
+State machine intent (currently no mandate to enforce)
+
+- Received -> in_progress or canceled.
```
Comment (philvarner) [inline opportunity/README.md:221]:
```suggestion
- received -> in_progress or canceled.
```

File: opportunity/README.md [hunk 17]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+entity defined to model the required search state.
+
+### Opportunity Search Record
+
+Returned by an async opportunity search. Can also be retrieved directly.
+
+| Field Name | Type | Description |
+| ---------- | ---- | ----------- |
+| id         | string | **REQUIRED.** Opportunity search record ID. |
+| productId  | string | **REQUIRED.** Product identifier. This should be a reference to the [Product](https://github.com/Element84/stapi-spec/blob/main/product/README.md) being searched. |
```
Comment (swallace-ursa) [inline opportunity/README.md:152]: For consistency, should this actually be `product_id`?

Comment (philvarner) [reply opportunity/README.md:152]: It should be `productId` because it's a reference to a path parameter and we use lowerCamelCase for those (don't ask me why we use lCC for those and snake_case for fields). But, I think this one needs to be separated out into a separate table for path params (even though this is the only one), as the rest of these are body entity fields.

File: opportunity/README.md [hunk 18]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+| status     | [Opportunity Search Status](#opportunity-search-status) | **REQUIRED.** The current search status. |
+| links      | [[Link Object](#opportunity-search-links)] | List of link objects to resources and related URLs. |
+
+#### Opportunity Search Links
```
Comment (jkeifer) [inline opportunity/README.md:156]: should call out self link

File: opportunity/README.md [hunk 19]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+
+### Opportunity Collection Links
+
+Each link in the links array must be a
+[Link](https://github.com/radiantearth/stac-spec/blob/master/commons/links.md#link-object)
+Object.
+
+#### Pagination links
+
+Links for pagination are required when pagination is supported and required.
```
Comment (jkeifer) [inline opportunity/README.md:54]: ?

File: opportunity/README.md [hunk 20]
```diff
@@ -8,77 +8,215 @@ The STAPI Opportunity describes a single business unit available for ordering.
...
+
+STAPI has an optional conformance class providing support for async opportunity
+searches, to accommodate searches for products that require more time to
+complete than is a typical or desirable HTTP request duration. This async
+support requires persisting the search state and matching opportunities outside
+the context of the initial request, so they are available for retrival at a
+later time by the requestor.  The Opportunity Search Record is an additional
+entity defined to model the required search state.
+
+### Opportunity Search Record
```
Comment (jkeifer) [inline opportunity/README.md:145]: Need to also include the search parameters themselves as part of the search object.

File: README.md (hunk 9 above)
Comment (jkeifer) [reply README.md:156]: Should be removed. `POST` endpoints are internal to the system and thus should be outside the scope of the spec, and up to the specific implementation to define.

File: opportunity/README.md (hunk 17 above)
Comment (jkeifer) [reply opportunity/README.md:152]: I believe this should be product_id, as @swallace-ursa suggests, as this is a property of the response object here, not a path parameter.

**Comment by matthewhanson** [comment]
@jkeifer to split search out from order and search record models

File: README.md [hunk 21]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
+order specific attributes and values in the feature properties.
+
+Implementations can choose to support synchronous or asynchronous opportunity
+search requests as needed to support the duration of the operations required to
+find opportunities for a given Product. In cases where the opportunity search
+can be completed within the duration of a regular HTTP request, synchronous
+operation is feasible and does not require maintaining any request or
+opportunity state outside the context of the request.
+
+Where opportunitiy search takes a longer duration, asynchronous operation
```
Comment (emmalu) [inline README.md:121]:
```suggestion
Where an opportunity search takes a longer duration, asynchronous operation
```

File: README.md [hunk 22]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
+| `POST /products/{productId}/opportunities` | Opportunities | [Opportunity Request](./opportunity/README.md#opportunity-request) | Sync search: [Opportunities Collection](./opportunity/README.md#opportunities-collection); Async search: [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | Explore the opportunities available for a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) |
+| `GET /products/{productId}/opportunities/{opportunityCollectionId}` | Opportunities (Async) | - | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Get the opportunity collection for an async opportunity search |
+| `GET /searches/opportunities` | Opportunities (Async) | - | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] | |
+| `GET /searches/opportunities/{searchRecordId}` | Opportunities (Async) | - | [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | |
```
Comment (emmalu) [inline README.md:154]:
```suggestion
| `GET /searches/opportunities/{searchRecordId}` | Opportunities (Async) | - | [Oportunity Search Record](./opportunity/README.md#oportunity-search-record) | |
```

File: README.md [hunk 23]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
+core ones detailed in the table below.
 
-The core STAPI conformance classes communicate the conformance JSON only in the root (`/`) document, while OGC API
-requires they also live at the `/conformance` endpoint. STAPI's conformance structure is detailed in the
-[core](core/). Note all conformance URIs serve up a rendered HTML version of the corresponding OpenAPI document at the given location.
+The core STAPI conformance classes communicate the conformance JSON only in the
+root (`/`) document, while OGC API requires they also live at the
+`/conformance` endpoint. STAPI's conformance structure is detailed in the
+[core](core/). Note all conformance URIs serve up a rendered HTML version of
+the corresponding OpenAPI document at the given location.
```
Comment (emmalu) [inline README.md:169]:
```suggestion
[core](core/). 
> [!NOTE]
> All conformance URIs serve up a rendered HTML version of
the corresponding OpenAPI document at the given location.
```

File: README.md [hunk 24]
```diff
@@ -211,7 +264,8 @@ sequenceDiagram
     deactivate PROVIDER
 ```
 
-A user with a specific product in mind views available opportunities and places and order.
+A user with a specific product in mind views available opportunities and places
+and order.
```
Comment (emmalu) [inline README.md:268]:
```suggestion
an order.
```

File: README.md [hunk 25]
```diff
@@ -245,3 +300,36 @@ sequenceDiagram
     PROVIDER-->>USER: Response: Order Object
     deactivate PROVIDER
 ```
+
+A user with a specific product in mind makes an async opportunities search
+request, retrieves the available opportunities, and places and order.
```
Comment (emmalu) [inline README.md:305]:
```suggestion
request, retrieves the available opportunities, and places an order.
```

File: README.md [hunk 26]
```diff
@@ -79,73 +97,106 @@ Fields that can be included in the response body for `GET /`.
...
-See [the STAPI Demo](https://github.com/Element84/stat-api-demo)
+See [the STAPI Demo](https://github.com/Element84/stat-api-demo).
 
 ## Pagination
 
 STAPI supports paging through hypermedia links for the following resources:
+
 - `GET /products`
 - `POST /products/{productId}/opportunities`
+- `GET /products/{productId}/opportunities/{collectionId}`
```
[the remark this answers is not in the document]
Comment (jkeifer) [reply README.md:193]: Good catch, it should be `/products/{productId}/opportunities/{opportunityCollectionId}`. I am updating it.

**Review by emmalu — APPROVED** [review]
🚀

File: README.md (hunk 23 above)
Comment (emmalu) [reply README.md:169]: No need to implement!

File: README.md [hunk 27]
```diff
@@ -245,3 +300,36 @@ sequenceDiagram
...
+    deactivate PROVIDER
+
+    loop Until search record is completed and has link to opportunity collection
+        USER->>PROVIDER: GET /searches/opportunities/{searchRecordId}
+        activate PROVIDER
+        PROVIDER-->>USER: Response: Opportunity Search Record
+        deactivate PROVIDER
+    end
+
+    USER->>PROVIDER: GET /products/{productId}/opportunities/{collectionId}
```
Comment (emmalu) [inline README.md:326]:
```suggestion
    USER->>PROVIDER: GET /products/{productId}/opportunities/{opportunityCollectionId}
```

**Review by trevorskaggs — APPROVED** [review]
🚢 🇮🇹
